Add Scaladoc tables

janekdb opened the pull request:
Parse a subset of GitHub flavoured markdown to define simple tables which can include inline wiki markdown. The specification for the table markdown is based on [GitHub Flavored Markdown Spec, Version 0.28-gfm (2017-08-01)](https://github.github.com/gfm/#tables-extension-).

The markdown used to structure the table follows [GitHub Flavoured Markdown Table Extension](https://github.github.com/gfm/#tables-extension-) with the restriction that pipe symbols at the start and end of row are mandatory and the clarification that delimiter row cells must have a minimum of three hyphens.

Note: Most of the following is irrelevant now the markdown is GHFM based..

```
  |     Define table cells
  ---   Default text alignment
  :---  Left text alignment
  ---:  Right text alignment
  :---: Center text alignment
```

Example.

A simple 2x2 table with a caption and header row,

```
  {|
  |+ Caption: Example Table
  ! header 1        ! header 2        !
  | row 1, column 1 | row 1, column 2 |
  | row 2, column 1 | row 2, column 2 |
  |}
```

Header and cell markdown can be repeated at the start of the heaer or row to allow | and ! to be included in content,

```
  {|
  !!  Disjunction!    !!  Conjunction!     !!
  ||| Logical OR = || ||| Logical AND = && |||
  |}
```

Fuller example including longer cell marks,

```
  {|
  |+ Caption: Example Table
  ! Header cell 1 ! Header cell 2 !
  | Row 1, column 1 | Row 1, column 2 |
  || Content with pipe symbol:  |A| := det(A) || final cell over
  two lines||
  |}
```

Caption, header and cell content can be broken over multiple lines and
include paragraphs and inline markdown,

```
    |Multiline
    cell
    content with __inline__ markdown

    Paragraph two.
```

Block level elements other than paragraphs are not yet parsed. Specifically these types are not yet supported,

 - Code blocks
 - Horizontal rules
 - Titles
 - Lists

A warning is issued when the rows and optional header do not all have the same number of cells.
This reserves an option to do something "wiki-ish" when cell counts differ.

Inspired by MediaWiki markdown with new row conciseness following GitHub flavoured markdown.

GitHub Project with subtasks: https://github.com/janekdb/scala/projects/1

## Live Demo

http://janekdb.github.io/scala/PR-6043/scala/test/scaladoc/tables/code/index.html

This page contains the unit test tables with the table markdown duplicated into a code block,

![image](https://user-images.githubusercontent.com/1123855/37338723-337e0696-26b0-11e8-8228-634415c99228.png)

I'm hoping this makes it easier to review the PR and also furnishes a good idea of what you get for what you give.

## PR Status

WIP

## Review Comments

 - [x] More compact way of introducing new rows

## TODO: Move to GHFM

 - [x] Make GHFM compliant
 - [x] Add GHFM column alignment options
 - [x] Update commit message
 - [x] Document requirement for delimiting pipes on rows per @lrytz comment
 - [x] Add negative test for missing pipes

## TODO: Support multiline content cell with GHFM extension

 - [ ] Understand dotty 3rd party markdown parser
 - [ ] Explore conversion of pre-dotty markdown to allow dotty markdown parser to be used
 - [ ] Retain multiline support subject to review of dotty approach

## TODO

 - [x] Include table start in check for new paragraph
 - [x] Fix lines numbers in `table-warnings.check`
 - [x] Add progress check in `parseCells0`
 - [x] Parse `Caption` as `Seq[Block]` matching `Cell`.
 - [x] Fix missing paragraphs in cell content starting with `CellsParagraphsB` and friends 
 - [x] Remove new row markdown `|-` because final cell termination is sufficient
 - [x] Consolidate or remove identically structured test traits
 - [x] CSS style table to blend into current Scaladoc themes not including row alternation styles
 - [x] Consider review comment regarding paragraph element remembering CSS is the correct way to address this
 - [x] Decide which other block elements other than Paragraph to parse: Code, Lists, Title, HRule.
 - [x] Review commit message
 - [x] Update grammar on table() method comment
 - [x] Remove `isNewlineAnyTableMarks` or the duplicate
 - [x] Run static analysis
 - [x] Squash commits

## WONT

Item | Reason | Comment
--- | --- | ---
Be more lenient toward leading whitespace below table marks|Deliver current tables sooner|Possible future improvement|
Support code blocks in cells in first merge|Deliver current tables sooner|Parsing code blocks will be the next improvement. I'd rather separate this off into new work because parsing code block while avoid code duplication could be quite a significant transformation requiring some recursion to be added at the top level of the WikiParser.|
Drop requirement for table start and end markers|Deliver current tables sooner|Without the start and end markers detecting transitions into and out of table parsing mode would be more complex and invasive. If ever we want some table level options around table formatting etc there would be no where to locate this information. Main reason is the implementation complexity.

## Acceptance Testing

 - [x] Check HTML rendering of test traits
 - [x] No regressions on existing non-table markdown in Scala library

## Review

 - [x] Create version of tables.scala which includes the table markdown as a code block and share it
 
## Target Outcome Over the Course of a few PRs :)

 - [ ] Tables as seen in ScalaTest docs can be created without requiring non-markdown work

## Follow Ons

- Support code blocks.
- Support other block elements: Horizontal Rules, Titles, List
- Make table css responsive
- Make styling in CaptionsB, CaptionsMultilineB effective instead of code style being bold and indistinguisable for other text
- Consider applying a minimum width for table to avoid caption wrapping as seen in CaptionsC
- Make styling of "and" in CellsC style the word differently to surrounding text.
- Left/right/centre cell alignment

janekdb commented:
This is an example of table markdown which includes all types of supported marks,

```
/** A variety of decorators that enable converting between
 *  Scala and Java collections using extension methods, `asScala` and `asJava`.
 *
 *  The extension methods return adapters for the corresponding API.
 *
 * {|
 * |+ Conversions are supported via `asScala` and `asJava`
 * !Scala                            !Direction!Java
 * |-
 * |scala.collection.Iterable        |<=>|java.lang.Iterable
 * |-
 * |scala.collection.Iterator       |<=>|java.util.Iterator
 * |-
 * |scala.collection.mutable.Buffer |<=>|java.util.List
 * |-
 * |scala.collection.mutable.Set    |<=>|java.util.Set
 * |-
 * |scala.collection.mutable.Map    |<=>|java.util.Map
 * |-
 * |scala.collection.concurrent.Map |<=>|java.util.concurrent.ConcurrentMap
 * |}
 *
```
![image](https://user-images.githubusercontent.com/1123855/29497758-de28e000-85e5-11e7-8bdd-95631f6169fb.png)

The unit test contains additional examples which include examples of multiline caption, header and cell content using inline wiki styling. Tables cannot contain other block elements.

At this point I've dealt with all the defects I could anticipate with the exception of one described below in *Known Issues*. I'l like to get this out to a wider audience now so the next set of defects can be found.

## Known Issues

Although it could be charitably regarded as adherent to the "spec" the markdown parsing does not tolerate leading whitespace before marks. To avoid wasting user time a subsequent PR will allow this to parse.

```
  // Should parse to table with a caption, header and a row
  /**
    * {|
    *  |+Leading
    *   whitespace before marks
    *    !!Not Yet Skipped!!TODO
    *    |-
    *    |A
    *       |B
    * |}
    */
  trait LeadingWhitespaceNotYetSkipped
```
Supporting this may require an extension to the `check*` method which so far I have managed to avoid. It would also require a few more unit tests and I feel since this limitation is well understood I'd rather get this in use and get some unanticipated cases reported.

## CSS Styling

The styling is a call to action for anyone with a passion for css. I will submit a PR if there is no interest from other quarters.

Follow on actions include

 - Converting some library Scaladoc to use tables where it makes an improvement
 - Updating the Scaladoc guide
 - Writing a blog post

adriaanm commented:
Cool! I'd prefer using GitHub-flavored markdown, as we do everywhere else. What do you think?

From https://guides.github.com/features/mastering-markdown/
> You can create tables by assembling a list of words and dividing them with hyphens - (for the first row), and then separating each column with a pipe |:

First Header | Second Header
------------ | -------------
Content from cell 1 | Content from cell 2
Content in the first column | Content in the second column

janekdb commented:
I surveyed a few markdown systems here: https://github.com/janekdb/scala/wiki/Table-Markdown-Survey

I believe the MediaWiki style markdown is easier to work with having used both GH and MW. It requires less attention to syntax than GH flavoured markdown and has the benefit of supporting vertical layout of cells in addition to horizontal layout which will work well of larger runs of inline markdown. MW can do everything GH can do AFAIK with less fiddling so I think it should be preferred.

adriaanm commented:
My opinion is that there's a strong case for standardization here, with feature completeness being a bit less urgent of a concern. Using GitHub markdown also makes it easier to preview

lrytz commented:
In general I agree to bias towards GFM, but here maybe the limitations are too severe?
  - GFM requires a header row
  - Cell content has to be on a single line
  - I like the having a caption, we'd have to invent something for GFM

Would it be OK adapt GFM to our needs?

On the other hand, the current proposal doesn't have functionality for left/center/right alignment.

I played a bit with the current state. One thing that's annoying for small tables is that each row requires a `|-` and a `|`. Example:

```
 * {|
 * |- | One | `1`
 * |- | Four | `4`
 * |}
```
It actually works without the first `|-`. I think it would be good if the `|` after `|-` was optional.

A detail in presentation: if two tables follow each other, the are sticked together. Maybe every table should be wrapped in a `<p>`?

The caption and header colors cold be a little less flashy :-)

I didn't get code blocks (`{{{ ... }}}`) to work within table cells - is that expected?

janekdb commented:
@lrytz —

 * Cell content is inline markdown only at this point - no block level elements.

 * The styling is deferred to a later commit.

I have some views to share regarding GHFM or not, just need a bit of time to write them down.

Thanks for the feedback. I am surprised the example was parsed but getting some feedback was the purpose of the PR!

lrytz commented:
OK, thanks for the clarifications!

I definitely think that allowing to use multiple lines for a cell's content (or even a row's) is very important. Otherwise the usefulness of this new feature would be severely limited.

I guess a good source of real-world examples is scalatest's scaladoc, searching for `<table>` in its checkout I find 42 results in 36 files. Ideally all of these should be doable in the new syntax.

Example: http://doc.scalatest.org/3.0.0/index.html#org.scalatest.AsyncFunSuite

janekdb commented:
@lrytz — The identification of these pre-existing table is super useful. This will allow for some concrete feature validation. I will take a look over the next few days.

adriaanm commented:
@janekdb, which version would you like to schedule this for? For 2.12.4, the PR must be mergeable by Wed Sep 27

janekdb commented:
@adriaanm — 2.12.5 as it needs rethinking to accommodate,

* More compact way of introducing new rows
* Tables as see in ScalaTest docs

janekdb commented:
/rebuild

janekdb commented:
Not forgotten. Making gradual progress on more compact row markdown.

janekdb committed:
Support GitHub Flavored Markdown version of tables in Scaladoc.

Based on GitHub Flavored Markdown Spec,

    https://github.github.com/gfm/#tables-extension-
    Version 0.28-gfm (2017-08-01)

A table is a block element consisting of,

  * A header row
  * A delimiter row separating the header from the data
  * Zero or more data rows

Restrictions,

    Rows must begin and end with pipe symbols
    A blank line required after table

Limitations,

    Escaping of pipe symbols is not yet supported

Inline markdown can be used in header and data cells, block markdown cannot be used.

Example,

    /**
      * |Nibbles|Main|Desert|
      * |:--:|:---:|----|
      * |Bread|Yak|Vodka|
      * |Figs|Cheese on toast^three ways^|Coffee|
      */
  trait RepastOptions

The accepted markdown is intended to be a strict subset of all possible GHFM tables.

janekdb commented:
Still working on a regression defect.

janekdb commented:
Regressions fixed. Clean up and then some end-to-end testing next. ☕️ 🍰

janekdb commented:
![scaladoc-table-java-convertors](https://user-images.githubusercontent.com/1123855/36377855-82b039b0-1570-11e8-8afc-fbe00190271a.gif)

GitHub flavour table row markdown is working.

Dreadful colour scheme is temporary!

Some more TODOs listed in the this issue's description before I ask for reviews but always happy to get some feedback now.

lrytz commented:
Happy to see you getting back to this!

SethTisue commented:
can't wait to see what @bvenners does with this

janekdb commented:
Note to self: Proof paragraphs are parsed in cells:

![image](https://user-images.githubusercontent.com/1123855/36784781-a05683a2-1c78-11e8-8c5b-6c9ab7eabd05.png)

janekdb commented:
I've decided to defer support for code blocks to a later PR.

The introduction of tables is mainly additive in the sense that only two existings methods in `CommentFactoryBase.scala` have been touched and then only to extend the cases for `table`. Everything just is achieved within the `table()` method.

Supporting code blocks and the other block elements within cells will need some reworking of `WikiParser` to avoid duplicating code inside the table parsing method. This might be accomplished by recursive parsing. Other block elements are horizontal rule, titles and lists. This work is very different to what has been done in adding tables so I want to cap the PR scope at this point.

With code block support as an immediate follower we won't be violating the principle of least surprise for too long.

Thoughts?

janekdb commented on src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala:
```diff
@@ -544,6 +562,376 @@ trait CommentFactoryBase { this: MemberLookupBase =>
+
+      def isNewline = char == endOfLine
+
+      def isNewlineAnyTableMarks = checkAny(TableAllMarksOnNextLine)
```
`isNewlineAnyTableMarks` is identical to `endOfInlineWhenTableMarkOnNextLine`. There can only be one `table-mark-on-next-line` incumbent!!!!

janekdb commented:
Tables are now spaced with a top and bottom `1em` margin.

![image](https://user-images.githubusercontent.com/1123855/36972465-0df7840a-2067-11e8-8758-b216d11ce1e3.png)

janekdb commented:
Table CSS has been subjected to a severe beating resulting in a new found and gladly embraced aversion to harmful individualistic tendencies.

![image](https://user-images.githubusercontent.com/1123855/37029891-d46a90d6-2130-11e8-8c63-88a6923a1be7.png)

janekdb commented:
This is ready for review. Please check the **Live Demo** section in the PR description for a useful link showing table markdown and the tables generated.

There are a number of very reasonable extensions to tables which I've listed in the description. I would prefer to follow these up in later PRs.

janekdb commented:
The regression test passed.
```
$ diff -r build-scaladoc-tables/scaladoc-quick/  build-scaladoc-2.12.x/scaladoc-quick/
diff -r build-scaladoc-tables/scaladoc-quick/lib/template.css build-scaladoc-2.12.x/scaladoc-quick/lib/template.css
822,847d821
< /* Author Content Table formatting */
< 
< .doctbl {
<   border-collapse: collapse;
<   margin: 1.0em 0em;
< }
< 
< table.doctbl caption {
<   text-align: center;
<   font-weight: bold;
<   font-size: 1.1em;
< }
< 
< table.doctbl th {
<   border: 1px dotted #364550;
<   background-color: #c2d2dc;
<   padding: 5px;
<   color: #103a51;
<   font-weight: bold;
< }
< 
< table.doctbl td {
<   border: 1px dotted #364550;
<   padding: 5px;
< }
< 
```
I build the Scaladoc for `library` using the `quick` binary from the `2.12.x` branch and this branch. The only differences as the expected new lines in `template.css`.

This PR is open for review!

janekdb commented:
/rebuild

lrytz commented:
@janekdb I'll take a look, but probably not this week..

janekdb commented:
Thanks @lrytz. I'll take your promissory note and bank it! Since this PR has been brewing for the average gestation period of an [American Bison](https://en.wikipedia.org/wiki/List_of_mammalian_gestation_durations) next week or even next is super early enough.

janekdb commented on src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala:
```diff
@@ -544,6 +562,370 @@ trait CommentFactoryBase { this: MemberLookupBase =>
+          storeContents()
+          skipNewline()
+        }
+        else if (isNewline) {
```
Fix new line

janekdb commented:
We are open 24-7 for all your reviewing comments needs!

lrytz commented:
Thanks @janekdb - I haven't forgotten, but got really busy with helping to integrate the 2.13 collections...

janekdb commented:
/rebuild

janekdb commented:
Build failure was: `The job exceeded the maximum time limit for jobs, and has been terminated`

lrytz commented:
I personally like this change and think it's good to go. Amazing work also on the testing front!

But I won't just decide on my own, because it has a long-term impact.

I guess the main concern people could have is the syntax. It is inspired by GitHub Markdown Tables (`| cell | cell | cell |`), but there are differences:
  - Cell content can be on multiple lines (GFM requires a single line per table row)
  - Support for captions
  - Table headers are different
  - `{|` / `|}` tokens for table start / end
  - plus some more

I asked on dotty's gitter (https://gitter.im/lampepfl/dotty?at=5b0db842b8abab25d37ea656). `dottydoc` supports scaladoc wiki, but also markdown. In fact it also supports some extensions by the flexmark-java implementation (https://github.com/vsch/flexmark-java/wiki/Tables-Extension). @smarter and @olafurpg argue that supporting plain markdown and deferring to html tables for more complex cases would be a better trade-off.

The drawback of falling back to html is that wiki syntax is gone entirely, so you end up writing a lot of html (`<a>`, `<li>`, `<code>`) withing tables, which I dislike.

cc @adriaanm, @SethTisue.

olafurpg commented:
Markdown + HTML have the benefit of being easier to reproduce in external tools like IntelliJ. I think custom Scaladoc markup syntax should ideally be limited to functionality that is not possible to encode in standard markup languages.

adriaanm commented:
So, to summarize the gitter discussion: the preference is to use Github flavored markdown (which means: drop support multi-line tables from this PR). 

I'm a big +1 on dropping any non-standard wiki syntax and standardizing on github's flavor of markdown. In the mean time, this PR has been in the making for a long time, so I propose we merge as-is, and start the migration to pure GFM as a separate work item. Plenty of other changes will be required.

lrytz commented:
I wouldn't release someting in 2.12.7 and change it again in 2.12.8 though. If the end goal is GFM only, then we should not release additional features (optional header row, captions, multi-line cells) just to remove them again.

adriaanm commented:
True, my thinking was that the expected total pain is low (== usage of
multi-line table extension * pain to convert to html) compared to the
upside. Ideally, the multiline extension should be disabled in this PR,
assuming that’s the only delta compared to GFM
On Thu, May 31, 2018 at 12:37 Lukas Rytz <notifications@github.com> wrote:

> I wouldn't release someting in 2.12.7 and change it again in 2.12.8
> though. If the end goal is GFM only, then we should not release additional
> features (optional header row, captions, multi-line cells) just to remove
> them again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala/pull/6043#issuecomment-393489595>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFjyzrcqsIyrmkuLvM57JDG6ZXP-O50ks5t38fggaJpZM4O8s5o>
> .
>

janekdb commented:
Thanks for feedback. @lrytz It was a nice chance to do some pure TDD!

Based on the observations this is the plan,

- Reduce differences to GHFM starting with dropping table open/close and reviewing header definitions
- Retain support for multiline content and make the case for this variation

lrytz commented:
Let's all agree on what to do before starting more work.

If I understand correctly, @adriaanm would really like GHFM only, which means
  - no captions
  - no multi-line contents
  - mandatory header row
  - (slightly) more verbose syntax than the current PR
  - no `||` cell separators, `\|` escape for pipe character inside a cell
  -add support for GHFM's column alignments (could come in a later PR)

janekdb commented:
Good to start the process of converging on agreed goals before restarting this.

Multiline content is important. It paves the way for allowing code blocks in cell content. With this in mind I assign limited importance to including the main deficiency of GHFM which is lack of support for multiline content other than via embedding `<br>`s. This can result in long lines that become inconvenient to edit which is contrary to the spirit of markdown. Some organisations use coding guidelines that impose a limit on line length.

Bailing out to HTML table tags when a cell needs to now have two paragraphs instead of one is fairly distracting and wasteful from an author's POV. With adequate markdown we can support the user in following the guidance to avoid HTML given on https://docs.scala-lang.org/style/scaladoc.html.

Let's not forget that GHFM is _currently_ a popular markdown variation but that status is no doubt a passing thing. My preference is to judge the solution by the intrinsic value. In this case support for multiline content outweighs the surprise factor.

Finally let's stand back: GHFM may already be an outgoing de facto standard. We could all be using [GitLab Flavored Markdown](https://docs.gitlab.com/ee/user/markdown.html) in the not too distant future.

lrytz commented:
I agree with, but it seems not everyone does. We need to decide somehow. Since this is scala/scala, @adriaanm could take the decision if he wants, otherwise we can take it to a SIP meeting.

adriaanm commented:
I don't think we need to SIP this. I would like compliance with GHFM as the default policy, but we can always discuss case by case where it makes sense to deviate. Supporting multiline content for tables is fine with me.

lrytz commented:
Cool! I agree to stick to GHFM and have extensions where we really think it adds significant and necessary value.

An issue is that dotty currently uses an off-the-shelve markdown parser, which wouldn't work anymore.

adriaanm commented:
Could we use some kind of pre-processing step for our extensions, and use the same markdown parser as dotty (or all agree on using one that perhaps supports the desired extensions?)

janekdb commented:
@adriaanm I'm happy to explore that before making changes here.

janekdb commented:
Would it be acceptable to implement a subset of the GHFM table syntax rules?

Namely not implementing the optional requirement on pipes at the start or end of a table line.

The reason for asking is if we require pipes on either side of every cell the parsing code would be somewhat different to the initial more expressive version that allows captions and multiline cells.

If leading and trailing pipes are supported then the parsing will need to change from the current modal greedy approach where the lookahead window is a few consecutive characters. The lookahead would need to be non-local to check for the existence a full table structure. This can be done efficiently by creating a suitable view of the raw markdown once at the beginning of the parse. I guess I've just about answered my own question in the stating of it!

Here is a minimal GHFM table according to the [spec](https://help.github.com/articles/organizing-information-with-tables/),

```
Command
---
git status
git diff
```

## Aside

Interestingly this legal table definition does not render correctly as show below,

_Valid GHFM Example 1: start_

Command
---
git status
git diff

_Valid GHFM Example 1: end_

However invalid GHFM markdown does render. The requirement for the header line to have at least three hyphens is skipped yet it renders,
```
Command
-|
git status
git diff
```

_Invalid GHFM Example 1: start_

Command
-|
git status
git diff

_Invalid GHFM Example 1: end_

Command
-|
git status
git diff

It won't be practical to match the vagaries of Github implementation of their spec so I would need to proceed by following the spec alone.

Comments?

janekdb commented:
The conversion to GHFM is progressing surprisingly straightforwardly now that we are working with the subset of all possible GHFM tables that have pipes at the start and end of each row. While I am sad face to see captions being removed we are gaining column alignment options.

adriaanm commented:
@janekdb The 2.12.7 window is closing next week (we may extend for max one week as needed). It sounds like this PR could make that release. What do you think?

janekdb commented:
@adriaanm Happy paths are going fine so Sept 14 is not impossible.

The main outstanding areas are,

1. Handle escaping of | with \\|
2. Ensure all multiline line support is removed
3. Tidy up

I'd have to defer work on negative cases and regard those as bug fixes. Would it be acceptable to also defer work on escaping | and get that in a bugfix?

lrytz commented:
Fixing these things in follow-ups sounds good to me too, sure!

janekdb commented:
Okay, good. I will concentrate on 2 & 3

- Ensure all multiline line support is removed
- Tidy up including squashing commits

lrytz commented:
@janekdb 
```
[test] [error]   - reflect/compile:doc failed: scala.NotImplementedError: an implementation is missing
[test] [error]   - compiler/compile:doc failed: scala.NotImplementedError: an implementation is missing
```

and

```
[test] !! 66 - run/tables-warnings.scala                 [compilation failed]
```

janekdb commented:
Seeing some encouraging HTML now...
![image](https://user-images.githubusercontent.com/1123855/45511036-71a86a80-b78b-11e8-98e1-7455e9f1a415.png)

lrytz commented on src/scaladoc/scala/tools/nsc/doc/base/CommentFactoryBase.scala:
```diff
@@ -544,6 +568,340 @@ trait CommentFactoryBase { this: MemberLookupBase =>
+
+        val inlineTerminators = makeInlineTerminators(startMark)
+
+        val content = Paragraph(inline(isInlineEnd = checkAny(inlineTerminators)))
```
Is it true that all terminators are single chars? Then this could be made more efficient by just checking the buffer's current `char` instead of calling `check` many times. (We can do this in a follow-up PR). A first simple improvement would be to test in `check` if the `chars` string has length 1, and avoid calling `jump` in that case.

janekdb replied:
I'm doing the clean up now so will look at this.

lrytz reviewed (commented):
LGTM! This is good to go once you're done with squashing and cleaning up.

janekdb commented:
@lrytz - This is the merge candidate.

Last actions,
* Removed dead code
* Actioned some TODOs, consolidated remaining TODO comments
* Fixed incorrect parsing of empty leading cell
* Investigated your optimization suggestion to use `Char` checks now that the only delimiter is the pipe. It will work but not with the time I have today.
 * Squashed commits
 * Ran end-to-end test

Not done,
 * Make scala project Scaladoc
 * Make community Scaladoc

Every item of suboptimal parsing has a test case. This will make it easy to continue with the edge cases following merge.

Compared to the per-GHFM version this had only light testing, however in balance the parsing code is much simpler thanks to the stricter requirements. The community build @SethTisue will give us more information about how well this copes with existing markdown.

I lay my offering at your collective feet.

adriaanm commented:
We gratefully and humbly accept your offerings! I have stood (by the sidelines) in awe at your diligence.

lrytz reviewed (approved):
🚀

janek-inferess commented:
Phew! Thanks for everyone's help and patience. I'll remind myself to get early community feedback before I plot a course the next time.

janekdb commented:
Phew! Thanks for everyone's help and patience. I'll remind myself to get early community feedback before I next plot a course.

SethTisue commented:
scala/community-builds@69bca5e2971e2d393476b5203fb42d65ac847644 moves to a Scala SHA that includes this PR, so we'll know by tomorrow if anything blows up

SethTisue commented:
@janekdb I would like to tweet about this feature from https://twitter.com/scala_lang (and credit you by name when we do so). do you have a tweet about it I could quote-tweet? (and/or could you propose a screenshot to include?)

janekdb commented:
@SethTisue How about this screenshot lifted from the forthcoming [blog post](https://github.com/janekdb/scala-lang/commits/topic/scaladoc-tables-part-one)

![image](https://user-images.githubusercontent.com/1123855/46226232-24abc300-c34b-11e8-8bc7-015900cbd39e.png)

You can crop off the intitial text leaving just the markdown and the table if you like.

My tweet: https://twitter.com/janekdb/status/1045742653822955520

janekdb commented:
@SethTisue Here's a couple of alternative images,

![image](https://user-images.githubusercontent.com/1123855/46284095-21405380-c566-11e8-83aa-ad2163ab7bcf.png)

![image](https://user-images.githubusercontent.com/1123855/46284227-8ac06200-c566-11e8-8c45-5a650364aac0.png)

SethTisue commented:
@janekdb I forgot about your blog post draft. rather than tweeting now, how about we publish the blog post and publicize the feature that way (including on Twitter)?

janekdb commented:
@SethTisue Blog post is done pending review: https://github.com/scala/scala-lang/pull/959
